### PR TITLE
Start adding a "tags" package.

### PR DESCRIPTION
**Pull request by sebright**
This commit adds TagKey and TagSet.  The classes are not used yet.

_______________________________
This change is part of #251.

/cc @dinooliva @songy23 @bogdandrutu

### CODE REVIEW

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @param <TagValueT> The type of value that can be paired with this {@code TagKey}.
+ */
+@Immutable
+@AutoValue
+public abstract class TagKey<TagValueT> {
+  /** The maximum length for a tag key name. */
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  enum TagType {
```
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Isn't the type encoded into the generic type?

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import static com.google.instrumentation.tags.TagKey.TagType.TAG_BOOL;
+import static com.google.instrumentation.tags.TagKey.TagType.TAG_INT;
+import static com.google.instrumentation.tags.TagKey.TagType.TAG_STRING;
+
+import com.google.auto.value.AutoValue;
+import com.google.instrumentation.internal.StringUtil;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Tag key.
```
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Please explain more about where is used and how is used. The current comment does not add any value.

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.tags;
+
+import com.google.instrumentation.internal.StringUtil;
+import javax.annotation.Nullable;
+
+/** A set of tags. */
```
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: What is a tag?

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+   * @param key the key to look up.
+   * @return {@code true} if the key is present.
+   */
+  public abstract boolean tagKeyExists(TagKey<?> key);
+
+  // TODO(sebright): Which is better, the generic "get" method or three specialized "get" methods?
+  // Should these methods take defaults or throw exceptions to handle missing keys?
+
+  /**
+   * Gets a value of type {@code String}.
```
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: s/Gets/Returns

Maybe change everywhere in this commit?

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    public abstract Builder insert(TagKey<String> key, String value);
+
+    /**
+     * Adds the key/value pair if the key is not present. If the key is present, it logs an error.
+     *
+     * @param key the key to look up.
+     * @param value the value to insert for the given key.
+     * @return this
+     */
+    public abstract Builder insert(TagKey<Long> key, long value);
```
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: We should follow java map or set (probably map) standards and use put or add instead.

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java (hunk 2 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: I wasn't sure what to write, but I tried to add an explanation in the TagSet Javadoc.  Are the new comments accurate?

**Comment by sebright** [comment]
I have a few questions about the API, which are in TODOs in TagSet.

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java (hunk 5 above)
Comment (adriancole) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: some of these imply the caller remembering if something was there before or a potentially racy check. Seems laborious vs put semantics. If this labor is worthwhile, we should probably explain in the type level docs why it is.

Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: > some of these imply the caller remembering if something was there before or a potentially racy check.

What is wrong with a racy check?  It is done in the builder, which is already not thread-safe.

> Seems laborious vs put semantics.  If this labor is worthwhile, we should probably explain in the type level docs why it is.

"set" has the same behavior as "put".  I think that the others are just convenience methods to save the user from having to call tagKeyExists() when they want to do something that depends on the existing value of a key.

/cc @a-veitch

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ * A key to a value stored in a {@link TagSet}.
+ *
+ * @param <TagValueT> The type of value that can be paired with this {@code TagKey}. {@code TagKey}s
+ *     can only be instantiated with types {@code String}, {@code Long}, and {@code Boolean}.
+ */
+@Immutable
+@AutoValue
+public abstract class TagKey<TagValueT> {
+  /** The maximum length for a tag key name. */
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: I suggest making the max a much smaller number for initial purposes - maybe 255?

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  TagKey() {}
+
+  /**
+   * Constructs a {@code TagKey<String>} from the given string. The string will be sanitized such
+   * that:
+   *
+   * <ol>
+   *   <li>length is restricted to {@link #MAX_LENGTH}, strings longer than that will be truncated.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: We should discuss what policies we want here. I think it would be better to raise an exception, rather than silently truncate or replace characters.

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+   *   <li>characters are restricted to printable ascii characters, non-printable characters will be
+   *       replaced by an underscore '_'.
+   * </ol>
+   */
+  public static TagKey<Boolean> createBoolean(String key) {
+    return new AutoValue_TagKey<Boolean>(StringUtil.sanitize(key), TAG_BOOL);
+  }
+
+  abstract String getName();
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Don't need either getName or getTagType in the v1 API.

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+/**
+ * A set of key-value pairs that can be used to label anything that is associated with a specific
+ * operation.
+ *
+ * <p>For example, {@code TagSet}s can be used to label stats, log messages, or debugging
+ * information.
+ */
+public abstract class TagSet {
+  /** The maximum length for a string tag value. */
+  public static final int MAX_STRING_LENGTH = StringUtil.MAX_LENGTH;
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: Why is this here? Raises potential for conflict with the same constant in TagKey. Suggest having this in one place only (probably TagKey).

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+  /**
+   * Determines whether a key is present.
+   *
+   * @param key the key to look up.
+   * @return {@code true} if the key is present.
+   */
+  public abstract boolean tagKeyExists(TagKey<?> key);
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: This implies that only one TagKey of a given name can exist at a time, which is not the case (you can have two tags with same key name, but different types). Need to match both name and type.

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+  // TODO(sebright): Which is better, the generic "get" method or three specialized "get" methods?
+  // The specialized "get" methods can return primitives, which could be less expensive.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: I think specialization is better, but don't feel strongly about it.

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+  // TODO(sebright): Should these methods take defaults or throw exceptions to handle missing keys?
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: The trouble with default values is that there may not be one - especially for bool. I think that leaves your only other option as being throwing; if users are want to avoid that, they can use tagKeyExists first.

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java (hunk 11 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: Thanks.  I removed the generic "get" method.  The set/update/insert methods were already specialized.

File: core/src/main/java/com/google/instrumentation/tags/TagSet.java (hunk 12 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagSet.java]: I removed the default parameters.  I also made getStringTagValue throw an exception if the key is missing, for consistency.

File: core_impl/src/test/java/com/google/instrumentation/tags/TagSetTest.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link TagSet}. */
+// TODO(sebright): Add more tests once the API is finalized.
+@RunWith(JUnit4.class)
+public class TagSetTest {
+
+  private static final TagKey<String> KS1 = TagKey.createString("k1");
+  private static final TagKey<String> KS2 = TagKey.createString("k2");
+  private static final TagKey<Long> KI = TagKey.createInt("k2");
```
Comment (a-veitch) [inline core_impl/src/test/java/com/google/instrumentation/tags/TagSetTest.java]: Should add a test doing something similar to (same key name, different types):

newBuilder().set(KS2, "v1").set(KI, 5).build()

File: core_impl/src/main/java/com/google/instrumentation/internal/logging/Logger.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
+/*
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/internal/logging/Logger.java]: I understand after reading further - log on errors. I think we can do away with this for now, since we have defined error-less operations for v1, and even when we do have them, we will (probably) want to use exceptions instead.

File: core_impl/src/test/java/com/google/instrumentation/tags/TagSetTest.java (hunk 13 above)
Comment (sebright) [inline core_impl/src/test/java/com/google/instrumentation/tags/TagSetTest.java]: Good catch!

**Comment by sebright** [comment]
I rebased onto master and updated error handling for invalid tag key names.  Now the public TagKey factory methods throw IllegalArgumentException when the name is invalid.

**Comment by sebright** [comment]
I temporarily removed all TagMap methods, except for `set` and `clear`, since it looks like we will definitely need those operations.  I can add other operations back in a separate PR, once we decide what we need.  This PR is ready for more review.

File: core/src/main/java/com/google/instrumentation/tags/TagMap.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * Adds the key/value pair regardless of whether the key is present.
+     *
+     * @param key the key to look up.
+     * @param value the value to set for the given key.
+     * @return this
+     */
+    public Builder set(TagKey<String> key, String value) {
+      Preconditions.checkArgument(key.getTagType() == TagType.TAG_STRING);
+      Preconditions.checkArgument(StringUtil.isValid(value));
```
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagMap.java]: Should we add a class to represent a string tag value, so that we can validate the string once and use it many times?

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+  TagKey() {}
+
+  /**
+   * Constructs a {@code TagKey<String>} from the given string. The string must meet the following
+   * requirements:
+   *
+   * <ol>
+   *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+   *   <li>It can only contain printable ascii characters.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Nit: upper-case ASCII

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+   *
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<String> createString(String name) {
+    Preconditions.checkArgument(StringUtil.isValid(name));
+    return createStringInternal(name);
+  }
+
+  // Creates a TagKey<String> by sanitizing the given String.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Not sure at all we want/need this functionality.

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+   *
+   * <ol>
+   *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+   *   <li>It can only contain printable ascii characters.
+   * </ol>
+   *
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<Long> createInt(String name) {
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: I think the name should be createLong as well - otherwise it's confusing.

File: core/src/main/java/com/google/instrumentation/tags/TagMap.java [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    Builder() {
+      this.tags = new HashMap<TagKey<?>, Object>();
+    }
+
+    /**
+     * Adds the key/value pair regardless of whether the key is present.
+     *
+     * @param key the key to look up.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagMap.java]: Suggest: @param key the {@code TagKey} which will be set

File: core/src/main/java/com/google/instrumentation/tags/TagMap.java [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+     * @param value the value to set for the given key.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagMap.java]: Need to add @throws documentation.

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java (hunk 18 above)
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Also other names, e.g. s/TAG_INT/TAG_LONG/

File: core/src/main/java/com/google/instrumentation/tags/TagMap.java (hunk 15 above)
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/tags/TagMap.java]: That would be fine as an optimization. Up to you whether you want to do this now, or add a todo.

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java (hunk 16 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Done

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java (hunk 17 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: I removed it.

File: core/src/main/java/com/google/instrumentation/tags/TagKey.java (hunk 18 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagKey.java]: Done

File: core/src/main/java/com/google/instrumentation/tags/TagMap.java (hunk 15 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagMap.java]: Thanks.  I added a TODO.

File: core/src/main/java/com/google/instrumentation/tags/TagMap.java (hunk 19 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagMap.java]: Done

File: core/src/main/java/com/google/instrumentation/tags/TagMap.java (hunk 20 above)
Comment (sebright) [inline core/src/main/java/com/google/instrumentation/tags/TagMap.java]: Done

**Comment by sebright** [comment]
I renamed the `tags` package to `io.opencensus.tags`.  The name isn't finalized, but it makes it clear that the package is part of the new API.

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+ * @param <TagValueT> The type of value that can be paired with this {@code TagKey}. {@code TagKey}s
+ *     can only be instantiated with types {@code String}, {@code Long}, and {@code Boolean}.
+ */
+@Immutable
+@AutoValue
+public abstract class TagKey<TagValueT> {
+  /** The maximum length for a tag key name. */
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  enum TagType {
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: I thought we are going to support only String type initially.

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+   *
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<String> createString(String name) {
+    Preconditions.checkArgument(StringUtil.isValid(name));
+    return createStringInternal(name);
+  }
+
+  private static TagKey<String> createStringInternal(String name) {
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: Why do you need this extra private method?

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+  enum TagType {
+    TAG_STRING,
+    TAG_LONG,
+    TAG_BOOLEAN
+  }
+
+  TagKey() {}
+
+  /**
+   * Constructs a {@code TagKey<String>} from the given string. The string must meet the following
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: The comment is confusing "from the given string". Probably better to say something like "with the given {@code name}" or something similar.

Also I would move the second sentence on a different line:
<p>The string must meet the following requirements:

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+   *
+   * <ol>
+   *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+   *   <li>It can only contain printable ASCII characters.
+   * </ol>
+   *
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<String> createString(String name) {
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: s/createString/createStringKey?

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * A key to a value stored in a {@link TagMap}.
+ *
+ * @param <TagValueT> The type of value that can be paired with this {@code TagKey}. {@code TagKey}s
+ *     can only be instantiated with types {@code String}, {@code Long}, and {@code Boolean}.
+ */
+@Immutable
+@AutoValue
+public abstract class TagKey<TagValueT> {
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: The current implementation defines equals/hashCode <name, type>. I think we haven't decided on this but you should be aware of this.

File: core/src/test/java/io/opencensus/tags/TagKeyTest.java [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+
+/** Tests for {@link TagKey} */
+@RunWith(JUnit4.class)
+public final class TagKeyTest {
+
+  @Rule public final ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testGetName() {
+    assertThat(TagKey.createString("foo").getName()).isEqualTo("foo");
```
Comment (bogdandrutu) [inline core/src/test/java/io/opencensus/tags/TagKeyTest.java]: Maybe a test for getType?

File: core/src/main/java/io/opencensus/tags/TagKey.java (hunk 25 above)
Comment (sebright) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: Yes, it shouldn't matter while we are only allowing String tags.

File: core/src/main/java/io/opencensus/tags/TagKey.java (hunk 21 above)
Comment (sebright) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: I made the factory methods for `TagKey<Long>` and `TagKey<Boolean>` non-public.

File: core/src/main/java/io/opencensus/tags/TagKey.java (hunk 23 above)
Comment (sebright) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: Done

File: core/src/main/java/io/opencensus/tags/TagKey.java (hunk 24 above)
Comment (sebright) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: Done.  I was trying to keep the name short, but createStringKey isn't much longer.

File: core/src/main/java/io/opencensus/tags/TagKey.java (hunk 22 above)
Comment (sebright) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: Previously, there were two factory methods for each type of key, but the private methods aren't needed now.

File: core/src/test/java/io/opencensus/tags/TagKeyTest.java (hunk 26 above)
Comment (sebright) [inline core/src/test/java/io/opencensus/tags/TagKeyTest.java]: Done

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+   * <ol>
+   *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+   *   <li>It can only contain printable ASCII characters.
+   * </ol>
+   *
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  public static TagKey<String> createStringKey(String name) {
+    Preconditions.checkArgument(StringUtil.isValid(name));
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: nit: Why not static import checkArguments? This is how we do everywhere else.

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+    TAG_LONG,
+    TAG_BOOLEAN
+  }
+
+  TagKey() {}
+
+  /**
+   * Constructs a {@code TagKey<String>} with the given name.
+   *
+   * <p>The string must meet the following requirements:
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: s/string/{@code name}

Probably everywhere else.

File: core/src/main/java/io/opencensus/tags/TagMap.java [hunk 29]
```diff
@@ -0,0 +1,137 @@
...
+      return this;
+    }
+
+    /**
+     * Creates a {@code TagMap} from this builder.
+     *
+     * @return a {@code TagMap} with the same tags as this builder.
+     */
+    public TagMap build() {
+      return new TagMap(new HashMap<TagKey<?>, Object>(tags));
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagMap.java]: collections.unmodifiablemap to avoid any mistake?

File: core/src/main/java/io/opencensus/tags/TagMap.java [hunk 30]
```diff
@@ -0,0 +1,137 @@
...
+     * @param value the value to set for the given key.
+     * @return this
+     * @throws IllegalArgumentException if either argument is null, the key is the wrong type, or
+     *     the value contains unprintable characters or is longer than {@link
+     *     TagMap#MAX_STRING_LENGTH}.
+     */
+    public Builder set(TagKey<String> key, String value) {
+      Preconditions.checkArgument(key.getTagType() == TagType.TAG_STRING);
+
+      // TODO(sebright): Consider adding a TagValue class to avoid validating the String every time
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagMap.java]: I think this is a good requirement to the design.

File: core/src/main/java/io/opencensus/tags/TagMap.java [hunk 31]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * Adds the key/value pair regardless of whether the key is present.
+     *
+     * @param key the {@code TagKey} which will be set.
+     * @param value the value to set for the given key.
+     * @return this
+     * @throws IllegalArgumentException if the key is null or the key is the wrong type.
+     */
+    public Builder set(TagKey<Boolean> key, boolean value) {
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagMap.java]: not public please.

File: core/src/main/java/io/opencensus/tags/TagMap.java [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * Adds the key/value pair regardless of whether the key is present.
+     *
+     * @param key the {@code TagKey} which will be set.
+     * @param value the value to set for the given key.
+     * @return this
+     * @throws IllegalArgumentException if the key is null or the key is the wrong type.
+     */
+    public Builder set(TagKey<Long> key, long value) {
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagMap.java]: not public please.

File: core/src/main/java/io/opencensus/tags/TagMap.java [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+   * Returns a builder based on this {@code TagMap}.
+   *
+   * @return a builder based on this {@code TagMap}.
+   */
+  public Builder toBuilder() {
+    return new Builder(getTags());
+  }
+
+  /** Builder for the {@link TagMap} class. */
+  public static final class Builder {
```
Comment (bogdandrutu) [inline core/src/main/java/io/opencensus/tags/TagMap.java]: At one point I thought that we will support something like:

public static TagMap build(TagMap parent, List<TagMutation>);

I think this way we can avoid 1 of the 2 copies of the tags.

Maybe not in this PR because we want to move forward but might worth playing with that idea.

**Comment by sebright** [comment]
Thanks!

File: core/src/main/java/io/opencensus/tags/TagMap.java (hunk 33 above)
Comment (sebright) [inline core/src/main/java/io/opencensus/tags/TagMap.java]: @dinooliva said that users probably won't add the same tag values every time, so it's probably not worth adding the `TagMap build(TagMap parent, List)` method.

File: core/src/main/java/io/opencensus/tags/TagKey.java [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+   * <ol>
+   *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+   *   <li>It can only contain printable ASCII characters.
+   * </ol>
+   *
+   * @param name the name of the key.
+   * @throws IllegalArgumentException if the name is not valid.
+   */
+  // TODO(sebright): Make this public once we support types other than String.
+  static TagKey<Long> createLongKey(String name) {
```
Comment (adriancole) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: to save future pull request/issue comments, I'd put a comment (not necessarily javadoc) rationale on choosing long vs float or something else to represent numbers. There are good reasons to choose Long, but in the future when people ask, this will help others who may not have made the decision help explain why.

Comment (sebright) [inline core/src/main/java/io/opencensus/tags/TagKey.java]: /cc @dinooliva

Comment (sebright) [reply core/src/main/java/io/opencensus/tags/TagKey.java]: I created an issue: #640
